### ISSUE TITLE
Adding homemade lmdb

### ISSUE THREAD

**Issue by LucMioulet** [NVIDIA/DIGITS#1109]
Hello,
I am trying to add a custom 64x64x3 lmdb image database for classification. This database was created using my own c++ code. So in fact I have 4 LMDB  : 
 -train_db 64x64x3
 -train_label_db 1x1x1
 -val_db 64x64x3
 -val_label_db 1x1x1
As it is requested by the "Other Classification Dataset".
At the moment, I am stuck on models that do not learn. 
Previously I tested creating the DB using the exact same images generated from my c++ code and using Digits to generate the lmdb via the gui. This worked very well.
Obviously I am doing something wrong … I just can't point where. Probably at datum level. I set the following : 
Datum datum, datumLabel;
datum.set_channels(num_channels);
datum.set_height(image_width);
datum.set_width(image_height);
cv::imencode(".png",mat_resized,buf1);
datum.set_data(std::string(reinterpret_cast<char*>(&buf1[0]), buf1.size()));
datum.set_encoded(true);

datumLabel.set_channels(1);
datumLabel.set_height(1);
datumLabel.set_width(1);
datumLabel.set_encoded(true);
if(classLabel==1){ 
                datum.set_label(1); //this is unused
                datumLabel.set_label(1);
                datumLabel.set_data("1");    
}
            else{ //Non pedestrian
                datum.set_label(0); 
                datumLabel.set_label(0);
                datumLabel.set_data("0");
}

My proto looks like this : 

# AlexNet
name: "AlexNet"
layer {
  name: "train_data"
  type: "Data"
  top: "data"
  data_param {
    backend: LMDB
    batch_size: 5
  }
  include: { phase: TRAIN }
}
layer {
  name: "train_label"
  type: "Data"
  top: "label"
  data_param {
    backend: LMDB
    batch_size: 5
  }
  include: { phase: TRAIN }
}
layer {
  name: "scale"
  type: "Power"
  bottom: "data"
  top: "scale"
  power_param {
    scale: 0.003921569
  }
}
layer {
  name: "conv1"
  type: "Convolution"
  bottom: "scale"
  top: "conv1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 96
    kernel_size: 11
    stride: 4
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "relu1"
  type: "ReLU"
  bottom: "conv1"
  top: "conv1"
}
layer {
  name: "norm1"
  type: "LRN"
  bottom: "conv1"
  top: "norm1"
  lrn_param {
    local_size: 5
    alpha: 0.0001
    beta: 0.75
  }
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "norm1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "conv2"
  type: "Convolution"
  bottom: "pool1"
  top: "conv2"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 256
    pad: 2
    kernel_size: 5
    group: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "relu2"
  type: "ReLU"
  bottom: "conv2"
  top: "conv2"
}
layer {
  name: "norm2"
  type: "LRN"
  bottom: "conv2"
  top: "norm2"
  lrn_param {
    local_size: 5
    alpha: 0.0001
    beta: 0.75
  }
}
layer {
  name: "pool2"
  type: "Pooling"
  bottom: "norm2"
  top: "pool2"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "conv3"
  type: "Convolution"
  bottom: "pool2"
  top: "conv3"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 384
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "relu3"
  type: "ReLU"
  bottom: "conv3"
  top: "conv3"
}
layer {
  name: "conv4"
  type: "Convolution"
  bottom: "conv3"
  top: "conv4"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 384
    pad: 1
    kernel_size: 3
    group: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "relu4"
  type: "ReLU"
  bottom: "conv4"
  top: "conv4"
}
layer {
  name: "conv5"
  type: "Convolution"
  bottom: "conv4"
  top: "conv5"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 256
    pad: 1
    kernel_size: 3
    group: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "relu5"
  type: "ReLU"
  bottom: "conv5"
  top: "conv5"
}
layer {
  name: "pool5"
  type: "Pooling"
  bottom: "conv5"
  top: "pool5"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "fc6"
  type: "InnerProduct"
  bottom: "pool5"
  top: "fc6"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "relu6"
  type: "ReLU"
  bottom: "fc6"
  top: "fc6"
}
layer {
  name: "drop6"
  type: "Dropout"
  bottom: "fc6"
  top: "fc6"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layer {
  name: "fc7"
  type: "InnerProduct"
  bottom: "fc6"
  top: "fc7"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "relu7"
  type: "ReLU"
  bottom: "fc7"
  top: "fc7"
}
layer {
  name: "drop7"
  type: "Dropout"
  bottom: "fc7"
  top: "fc7"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layer {
  name: "fc8"
  type: "InnerProduct"
  bottom: "fc7"
  top: "fc8"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    # Since num_output is unset, DIGITS will automatically set it to the
    #   number of classes in your dataset.
    # Uncomment this line to set it explicitly:
    num_output: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "accuracy"
  type: "Accuracy"
  bottom: "fc8"
  bottom: "label"
  top: "accuracy"
  include { stage: "val" }
  include { stage: "train" }
}
layer {
  name: "loss"
type: "SoftmaxWithLoss"
  bottom: "fc8"
  bottom: "label"
  top: "loss"
  exclude { stage: "deploy" }
  include { stage: "val" }
  include { stage: "train" }
}
layer {
  name: "softmax"
  type: "Softmax"
  bottom: "fc8"
  top: "softmax"
  include { stage: "deploy" }
}

Can anyone spot a mistake ? Thanks a lot.

**Comment by gheinrich** [comment]
Hello, what are you using a separate label DB (as opposed to the `label` field in `Datum` objects)?

**Comment by LucMioulet** [comment]
From what I understood, the label is unused in the image lmdb (see comment from @lukeyeager)  https://github.com/NVIDIA/DIGITS/issues/97.

**Comment by gheinrich** [comment]
The label field is unused if there is a dedicated label DB, otherwise it is used. I am not saying using only one DB will fix your issue but it will certainly limit the possibilities of doing anything wrong (e.g. not aligning the image DB with the label DB).

**Comment by LucMioulet** [comment]
Ok thanks for the tip ! I made some progress now. It still seems incorrect regarding the end results, but at least the loss is going down now. I will post the final solution once I found it.

**Comment by LucMioulet** [comment]
Here is the cpp code corrected :
Datum datum;
datum.set_channels(num_channels);
datum.set_height(image_width);
datum.set_width(image_height);
cv::imencode(".png",mat_resized,buf1);
datum.set_data(std::string(reinterpret_cast(&buf1[0]), buf1.size()));
datum.set_encoded(true);

if(classLabel==1){ 
datum.set_label(1);
}
else{
datum.set_label(0); 
}

However, my biggest grip was not there … in fact it was a bit after.

txn->Put(key, out);

I was using a string formated such as "/path/to/class/A" or "/path/to/class/B". When reading this file to retrieve an example, caffe uses the alphabetical order of the key. Hence explaining bad results.
I had to add a random string at the beginning of the key in order to shuffle the db.
